javascript/ql/src/Security/CWE-116/InvalidEntityTranscoding.qhelp

```diff
@@ -0,0 +1,71 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>
+Sanitizing untrusted input is an important technique for preventing injection attacks such as
+cross-site scripting. One particular example of this technique is HTML entity encoding, where
+HTML special characters in untrusted input are replaced by HTML character entities to prevent
+them from being interpreted as HTML markup. For example, the less-than character is encoded as
+<code>&amp;lt;</code> and the double-quote character as <code>&amp;quot;</code>. The reverse
+process of replacing character entities with the characters they represent is known as HTML
+entity decoding. Taken together, we refer to the process of encoding or decoding as
+<i>transcoding</i>.
+</p>
+<p>
+Note that the ampersand character plays a special role in HTML entity transcoding: it is itself
+encoded as <code>&amp;amp;</code>, but also forms part of the encoded representations of other
+characters. Hence care must be taken to avoid double transcoding: when encoding, the
+ampersand must be encoded first, when decoding it has to be decoded last.
+</p>
+<p>
+If the entity transcoding is used for sanitization, double transcoding may render the sanitization
+ineffective. Even if it is not used in a security-critical context, the double transcoding
+may still result in confusing or garbled output.
+</p>
+</overview>
+
+<recommendation>
+<p>
+Use a (well-tested) HTML entity transcoding library if at all possible. These libraries are much more
+likely to handle corner cases correctly than a custom implementation.
+</p>
+<p>
+Otherwise, make sure to always encode the ampersand first and decode it last.
+</p>
+</recommendation>
+
+<example>
+<p>
+The following example shows a pair of hand-written HTML encoding and decoding functions:
+</p>
+
+<sample src="examples/InvalidEntityTranscoding.js" />
+
+<p>
+The encoding function correctly handles ampersand before the other characters. For example,
+the string <code>me &amp; "you"</code> is encoded as <code>me &amp;amp; &amp;quot;you&amp;quot;</code>,
+and the string <code>&quot;</code> is encoded as <code>&amp;quot;</code>.
+</p>
+
+<p>
+The decoding function, however, incorrectly decodes <code>&amp;amp;</code> into <code>&amp;</code>
+before handling the other characters. So while it correctly decodes the first example above,
+it decodes the second example (<code>&amp;quot;</code>) to <code>&quot;</code> (a single double quote),
+which is not correct.
+</p>
+
+<p>
+Instead, the decoding function should decode ampersand last:
```

Comment:
Tiny suggestion:
Instead, the decoding function should decode THE ampersand last: